File: leet/0382_linked_list_random_node/0382_linked_list_random_node.js
// LeetCode 382. Linked List Random Node
// https://leetcode.com/problems/linked-list-random-node/
// O(n) time, O(1)
var Solution = function(head) {
    this.head = head;
};

Solution.prototype.getRandom = function() {
    let k = this.head, node = this.head.next, n = 2;
    while (node) {
        if (Math.floor(Math.random() * n) === 0)
            k = node;
        node = node.next;
        n++;
    }
    return k.val;
};

File: leet/0382_linked_list_random_node/0382_linked_list_random_node.cpp
// LeetCode 382. Linked List Random Node
// https://leetcode.com/problems/linked-list-random-node/
// O(N) time, O(1) space
class Solution {
public:
    ListNode* head;
    
    Solution(ListNode* head) {
        srand(time(NULL));
        this->head = head;
    }
    
    int getRandom() {
        int k = head->val, n = 2;
        for (ListNode *node = head->next; node; node = node->next) {
            if (!(rand() % n++))
                k = node->val;
        }
        return k;
    }
};

